### PR TITLE
first draft of a map<> explainer

### PR DESCRIPTION
**Pull request by mr-c**
Half of https://github.com/common-workflow-language/common-workflow-language/issues/529

### CODE REVIEW

**Comment by tetron** [comment]
* The language is a little breezy, but that's just a matter of personal preference
* Should mention the following pattern as a variation of "option two":
```
requirements:
  InlineJavascriptRequirement: {}
```

**Comment by tetron** [comment]
"hyperlink all usage of `map<>` to this explanation" would require schema salad support, since the "map<>" is generated?

**Comment by mr-c** [comment]
> The language is a little breezy, but that's just a matter of personal preference

Yes, it is purposely more in the style of the tutorial. We can mark it non-normative to be clear.

> Should mention the following pattern as a variation of "option two":

Good point! I'll add this tomorrow

**Comment by mr-c** [comment]
@tetron Done, ready for review & merge

File: Workflow.yml [hunk 1]
```diff
@@ -87,6 +87,7 @@ $graph:
         * [Added](#LoadListingRequirement) `LoadListingRequirement`
           and [loadListing](#LoadContents) to control whether and how
           `Directory` listings should be loaded for use in expressions.
+        * [Better explain](#map) the `map<…>` notation that has existing since v1.0.
```
Comment (tetron) [inline Workflow.yml]: typo "existed"

File: concepts.md [hunk 2]
```diff
@@ -91,6 +91,150 @@ An implementation may formally validate the structure of a CWL document using
...
+For this example, if we only need the `key_field` and `field2` when specifying
+our `ComplexType`s (because the other fields are optional and we are fine with
+their default values) then we can abbreviate.
+
+A generic example:
+```
+some_cwl_field:
+  a_complex_type1: foo   # we accept the default value for field3
+  a_complex_type2: foo2  # we accept the default value for field3
+  a_complex_type3: {}    # we accept the defualt values for "field2" and "field3"
```
Comment (tetron) [inline concepts.md]: typo "default"

File: concepts.md [hunk 3]
```diff
@@ -91,6 +91,150 @@ An implementation may formally validate the structure of a CWL document using
...
+### map
+
+Note: This section is non-normative.
+> type: array&lt;ComplexType&gt; |  
+> map&lt;`key_field`, ComplexType&gt;
+
+The above syntax in the CWL specifications means there are two or more ways to write the given value.
+
+Option one is a array and is the most verbose option.
+In our example here we use the generic`ComplexType`, but
```
Comment (tetron) [inline concepts.md]: typo missing space between generic`ComplexType`

File: concepts.md [hunk 4]
```diff
@@ -91,6 +91,150 @@ An implementation may formally validate the structure of a CWL document using
...
+in reality it would be one of `InputRecordField`, `OutputRecordField`,
```
Comment (tetron) [inline concepts.md]: Technically the mapping attaches to the field, not the types of the field, I'm not exactly sure if listing these out here is helpful enough to outweigh the burden of adding another place in the text that has to be maintained every time we update the spec.

File: concepts.md [hunk 5]
```diff
@@ -91,6 +91,150 @@ An implementation may formally validate the structure of a CWL document using
...
+
+
+```
+inputs:
+  - id: workflow_input01
+    type: string
+  - id: workflow_input02
+    type: File
+    format: http://edamontology.org/format_2572
+  - id: workflow_input03
```
Comment (tetron) [inline concepts.md]: The 3rd one is illegal, can't have inputs without a type

File: concepts.md [hunk 6]
```diff
@@ -91,6 +91,150 @@ An implementation may formally validate the structure of a CWL document using
...
+
+
+```
+inputs:
+  workflow_input01:
+    type: string
+  workflow_input02:
+    type: File
+    format: http://edamontology.org/format_2572
+  workflow_input03: {}  # we accept the default value for "type"
```
Comment (tetron) [inline concepts.md]: As noted above, there's no default value for `type` so maybe this should use a different example or take this part out.  The `{}` pattern most often shows up in `requirements`

**Review by tetron — CHANGES_REQUESTED** [review]

File: concepts.md (hunk 5 above)
Comment (mr-c) [reply concepts.md]: Not according to https://www.commonwl.org/v1.1.0-dev1/Workflow.html#WorkflowInputParameter "required: false"

**Pull request merged**